app/views/shared/_typeahead_search.html.erb

```diff
@@ -0,0 +1,82 @@
+<div class="typeahead">
+  <%= search_field_tag :search_query,
+                       nil,
+                       id: :search_query,
+                       class: "typeahead-input form-control mb-1",
+                       hide_label: true,
+                       placeholder: placeholder,
+                       onkeyup: "debouncedSearch(this)",
+                       autocomplete: "off" %>
+  <div class="typeahead-dropdown"></div>
+  <div class="typeahead-spinner">
+    <div class="spinner-border text-primary"></div>
+  </div>
+</div>
+
+<script>
+  function emptyTypeahead() {
+    $(".typeahead-dropdown").empty();
+    $(".typeahead-input").val("");
+  }
+
+  function resultsJSONToHTML(results) {
+    return results.map(function (result) {
+      return <%= search_result_to_row %>(result);
+    })
+  }
+
+  function noResultsHTML(searchQuery) {
+    return <%= no_results_message %>(searchQuery)
+  }
+
+  function searchResultsToHTML(searchQuery, results) {
+    return results.length ? resultsJSONToHTML(results) : noResultsHTML(searchQuery);
+  }
+
+  function populateDropdown(body) {
+    $(".typeahead .typeahead-dropdown").html(body);
+  }
+
+  function populateSearchResults(searchQuery, response) {
+    let filteredResults = <%= results_transformer %>(response);
+    let results = searchResultsToHTML(searchQuery, filteredResults);
+
+    populateDropdown(results);
+  }
+
+  function search(e) {
+    let searchQuery = e.value;
+    if (searchQuery.length) {
+      showSpinner();
+      $.ajax({
+        url: "<%= url %>",
+        data: {"search_query": searchQuery},
+        success: (res) => {
+          populateSearchResults(searchQuery, res)
+        }
+      })
+    }
+  }
+
+  function debounce(func, wait = 500) {
+    let timeout;
+    return function (...args) {
+      const later = () => {
+        clearTimeout(timeout);
+        func(...args);
+      };
+
+      clearTimeout(timeout);
+      timeout = setTimeout(later, wait);
+    };
+  }
+
+  let debouncedSearch = debounce(search);
+
+  function showSpinner() {
+    let spinner = $(".typeahead-spinner").first().clone();
+    spinner.css({display: "block"});
+
+    populateDropdown(spinner);
+  }
```

Comment:
We should put this JS in the JS pipeline as well.

Reply:
This JS needs to be in an erb file since there is some fancy rails templating going on. 
```
  function resultsJSONToHTML(results) {
    return results.map(function (result) {
      return <%= search_result_to_row %>(result);
    })
  }
```

This can be moved into separate `typeahead.js.erb`?

Reply:
Oh I see. Yeah that seems quite convoluted indeed 🤔  Could we leverage another template and handle the JSON results client-side, like we do for the already-added-user cards? This stuff needing to be in here is a good red flag that we may be able to clean up the implementation a bit 😅

Reply:
We are handling the JSON all on client side, this bit is just to support the typeahead as a reusable component. This allows us to pass in JS functions to the partial that will do the JSON to HTML business:
```
 <%= render 'shared/typeahead_search',
                       placeholder: "Find medical officers by name or phone number",
                       url: "/admin/users.json",
                       results_transformer: "filterExistingUsers",
                       search_result_to_row: "userJSONToRow"
...
```
This gives the user a lot of flexibility to structure/style the typeahead as they wish.

Reply:
Oh I see, hmm. I'd be tempted to hold off on the abstraction until we have a second, maybe even third, use case for typeahead. Keep it simple now, avoid guessing what the abstraction would look like without a concrete use case. What do you think?

Reply:
Yeah this seems like overkill, I'll revert back to the plebian version.